### PR TITLE
Changes for issue #19

### PR DESCRIPTION
**Pull request by tylerhawkes**
I ended up adding another feature which is the ability to add a suffix to the git version if the directory has any changes. 

Please squash all the commits on merge.

### CODE REVIEW

**Comment by lukaslueg** [comment]
Thanks for the PR! A few thoughts:

* Could we have the `dirty` flag as an `Option<String>` where `Option::is_some` (instead of the empty string) means the workspace was dirty? My original thought was to have simple `bool`, because the suffix can always be added by the downstream code itself.
* Could you add an integration test that demonstrates that git changes are picked up the way we expect them to? There is infrastructure for dogfood-tests in `/tests/testbox_tests.rs`.
* What about bare repositories? Those don't have the same path layout. Maybe `libgit2` can help out to figure out if the workspace is considered dirty (which includes handling `.gitignore` to that effect).

**Comment by tylerhawkes** [comment]
@lukaslueg Anything else you'd like me to do to get this merged?

File: src/lib.rs [hunk 1]
```diff
@@ -319,17 +322,20 @@ fn fmt_option_str<S: fmt::Display>(o: Option<S>) -> String {
...
 ) -> io::Result<()> {
     // CIs will do shallow clones of repositories, causing libgit2 to error
     // out. We try to detect if we are running on a CI and ignore the
     // error.
-    let tag = match util::get_repo_description(&manifest_location) {
-        Ok(tag) => tag,
-        Err(_) => None,
-    };
+    let tag =
+        match util::get_repo_description(&manifest_location, dirty_suffix, rerun_on_git_change) {
```
Comment (lukaslueg) [inline src/lib.rs]: This is just `util::get_repo_description(...).ok()` ?!

**Comment by lukaslueg** [comment]
The original code needs a little scrubbing anyway, I can do that afterward. My concern is with correctness in `get_repo_description()`. [Here](https://github.com/rust-lang/cargo/blob/dc83ead224d8622f748f507574e1448a28d8dcc7/src/cargo/ops/fix.rs#L129) is how `cargo` determines if a workspace is to be considered dirty, maybe we can check on that ?

I think we are good if the integration tests give a reasonable proof that this works as expected (e.g. a repo with a file changed that is being `.gitignore`d is not considered dirty); a bare repository works as expected. Shallow clones do not get their `tag` retrieved because `libgit2` may be unable to trace that; what's the behaviour of `dirty` then?

**Comment by tylerhawkes** [comment]
I'll create tests for all the things you brought up and investigate how it works with shallow clones. As far as the dirty flag goes, are you saying you'd like a new field output in built.rs? I can probably do that. The dirty flag is just an indication of whether or not there are tracked files that have changes to them. This is going to take a few days to get finished up though.

**Comment by lukaslueg** [comment]
Let me know if you need anything.

Yeah, a new flag in `built.rs`, so the downstream crate can have both `git tag/commit id` and `dirty`

**Comment by lukaslueg** [comment]
Oh btw, I didn't check: If `cargo` exports it's own "is this ws dirty?"-functionality as public functionality, it's ok to add `cargo` as a dependency.

**Comment by lukaslueg** [comment]
Are you still working on this?

**Comment by tylerhawkes** [comment]
Yes, it's still on my todo list, but I probably won't be able to get to it for another couple of weeks. Don't hold up any work on my account, I'll take care of any merges that need to happen.

**Comment by lukaslueg** [comment]
I've pushed a commit to the `dirty` branch which includes a new `GIT_DIRTY: Option<bool>` which tracks if the git repo had been dirty during build. Is this what you had in mind?

You can include this in your project by pointing the `built`-dependency to this repo#dirty instead of crates.io.

**Comment by tylerhawkes** [comment]
Yep. That's what I wanted, now I just need to get the tests done to get it to rebuild if the commit/tag it was on changes.

**Comment by lukaslueg** [comment]
I don't quite get what you mean. If the working directory changes and those changes are done to things rustc consume, the fingerprint will change and a rebuild is done automatically

**Comment by lukaslueg** [comment]
This is now implemented AFAICS

**Pull request closed without merging**